src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -284,90 +284,109 @@ static Solution UnifyLinkedDocumentContents(Solution oldSolution, Solution newSo
+                // not search the project that addedDocumentId came from.  So this will help ensure we don't repeatedly add
+                // documents to a project, then look for related docs *within that project*, forcing the file-path map in it
+                // to be recreated for each document.
+                var relatedDocumentId = solution.GetFirstRelatedDocumentId(addedDocumentId, relatedProjectIdHint);
+
+                // Couldn't find a related document.
+                if (relatedDocumentId is null)
+                    continue;
 
-            // Couldn't find a related document.  Keep the same solution, and keep track of the best related project we
-            // found while processing this project.
-            if (relatedDocumentId is null)
-                return (solution, relatedProjectIdHint);
+                var relatedDocument = solution.GetRequiredDocument(relatedDocumentId);
 
-            var relatedDocument = solution.GetRequiredDocument(relatedDocumentId);
+                // Should never return a file as its own related document
+                Contract.ThrowIfTrue(relatedDocumentId == addedDocumentId);
 
-            // Should never return a file as its own related document
-            Contract.ThrowIfTrue(relatedDocumentId == addedDocumentId);
+                // Related document must come from a distinct project.
+                Contract.ThrowIfTrue(relatedDocumentId.ProjectId == addedDocumentId.ProjectId);
+
+                relatedProjectIdHint = relatedDocumentId.ProjectId;
+                relatedDocumentIdsAndStates.Add((addedDocumentId, relatedDocument.DocumentState));
+            }
 
-            // Related document must come from a distinct project.
-            Contract.ThrowIfTrue(relatedDocumentId.ProjectId == addedDocumentId.ProjectId);
+            if (relatedDocumentIdsAndStates.IsEmpty)
+                return solution;
 
-            var newSolution = solution.WithDocumentContentsFrom(addedDocumentId, relatedDocument.DocumentState, forceEvenIfTreesWouldDiffer: false);
-            return (newSolution, relatedProjectId: relatedDocumentId.ProjectId);
+            return solution.WithDocumentContentsFrom(relatedDocumentIdsAndStates.ToImmutableAndClear(), forceEvenIfTreesWouldDiffer: false);
         }
 
-        static Solution UpdateExistingDocumentsToChangedDocumentContents(Solution solution, DocumentId changedDocumentId, HashSet<DocumentId> processedDocuments)
+        static Solution UpdateExistingDocumentsToChangedDocumentContents(Solution solution, HashSet<DocumentId> changedDocumentIds)
         {
             // Changing a document in a linked-doc-chain will end up producing N changed documents.  We only want to
             // process that chain once.
-            if (processedDocuments.Add(changedDocumentId))
+            using var _ = PooledDictionary<DocumentId, DocumentState>.GetInstance(out var documentIdsToUpdate);
```

Comment:
can this happen?  do we have any case where we'd see a doc-id twice for changed docs.

Reply:
I was being careful, but it seems like two docs could change that have the same related document, no?